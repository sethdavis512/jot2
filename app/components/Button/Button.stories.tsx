import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import Button from '.';

export default {
  title: 'Button',
  component: Button,
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  variant: 'primary',
  children: 'Primary',
};

export const Standard = Template.bind({});
Standard.args = {
  variant: 'standard',
  children: 'Standard',
};

export const Accent = Template.bind({});
Accent.args = {
  variant: 'accent',
  children: 'Accent',
};

export const Danger = Template.bind({});
Danger.args = {
  variant: 'danger',
  children: 'Danger',
};

export const Hollow = Template.bind({});
Hollow.args = {
  variant: 'hollow',
  children: 'Hollow',
};
