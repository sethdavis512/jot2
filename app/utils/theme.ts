import { Theme as StyledSystemTheme } from '@styled-system/css';

const makeTheme = <T extends StyledSystemTheme>(t: T) => t;
const theme = makeTheme({
  colors: {
    blue: {
      5: '#E8F5FF',
      10: '#CFE8FF',
      20: '#A1D3FF',
      30: '#58B4FF',
      40: '#2491FF',
      50: '#0076DA',
      60: '#005EA2',
      70: '#0B4778',
      80: '#112F4E',
    },
    gray: {
      0: '#FFFFFF',
      1: '#FBFCFD',
      2: '#F7F9FA',
      3: '#F5F6F7',
      4: '#F1F3F6',
      5: '#EDEFF0',
      10: '#DFE1E2',
      20: '#C6CACE',
      30: '#A9AEB1',
      40: '#8D9297',
      50: '#71767A',
      60: '#565C65',
      70: '#3D4551',
      80: '#2D2E2F',
      90: '#1C1D1F',
      100: '#000000',
    },
    mint: {
      5: '#D5FBF3',
      10: '#7EFBE1',
      20: '#29E1CB',
      30: '#1DC2AE',
      40: '#00A398',
      50: '#008480',
      60: '#0F6460',
      70: '#0B4B3F',
      80: '#123131',
    },
    red: {
      5: '#FFF3F2',
      10: '#FDE0DB',
      20: '#FDB8AE',
      30: '#FF8D7B',
      40: '#FB5A47',
      50: '#E52207',
      60: '#B50909',
      70: '#8B0A03',
      80: '#5C1111',
    },
  },
  fonts: {
    default: '"Source Sans Pro", "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif',
  },
  fontSizes: {
    1: '0.625rem',
    2: '0.75rem',
    3: '1rem',
    4: '1.25rem',
    5: '1.5rem',
    6: '2rem',
    7: '2.5rem',
    8: '3rem',
  },
  fontWeights: {
    normal: 400,
    bold: 700,
  },
  lineHeights: {
    1: '112.5%',
    2: '125%',
    3: '133.4%',
    4: '140%',
    5: '150%',
    6: '166.7%',
  },
  space: {
    1: '0.0625rem',
    2: '0.125rem',
    3: '0.25rem',
    4: '0.5rem',
    5: '0.75rem',
    6: '1rem',
    7: '1.25rem',
    8: '1.5rem',
    9: '2rem',
    10: '2.5rem',
    11: '3rem',
    12: '4rem',
    13: '4.5rem',
    14: '5rem',
  },
  radii: {
    1: '0rem',
    2: '0.125rem',
    3: '0.25rem',
    4: '0.5rem',
    5: '1rem',
    full: '100%',
  },
  styles: {
    root: {
      fontFamily: 'default',
      fontWeight: 'regular',
    },
  },
});

export enum modes {
  light = 'light',
  dark = 'dark',
}

export type Theme = typeof theme & {
  mode: modes;
};

export default theme;
