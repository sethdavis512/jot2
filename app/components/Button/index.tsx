import styled from '@emotion/styled';
import { themeHelper, sxPropHelper, StyleProps } from '../../utils/styled';
import { modes } from '../../utils/theme';

interface ButtonProps extends StyleProps {
  variant?: keyof typeof variants;
}

const variants = {
  standard: (mode: modes) =>
    mode === modes.dark
      ? {
          backgroundColor: 'gray.70',
          color: 'gray.5',
          ':hover': {
            backgroundColor: 'gray.60',
          },
          ':active': {
            backgroundColor: 'gray.100',
            borderColor: 'transparent',
          },
          ':focus-visible': {
            borderColor: 'blue.20',
          },
          ':disabled': {
            ':hover': {
              backgroundColor: 'gray.70',
            },
          },
        }
      : {
          backgroundColor: 'gray.5',
          color: 'gray.80',
          ':hover': {
            backgroundColor: 'gray.2',
          },
          ':active': {
            backgroundColor: 'gray.20',
            borderColor: 'transparent',
          },
          ':focus-visible': {
            borderColor: 'blue.50',
          },
          ':disabled': {
            ':hover': {
              backgroundColor: 'gray.5',
            },
          },
        },
  primary: (mode: modes) =>
    mode === modes.dark
      ? {
          backgroundColor: 'blue.50',
          color: 'gray.5',
          ':hover': {
            backgroundColor: 'blue.40',
          },
          ':active': {
            backgroundColor: 'blue.60',
            borderColor: 'transparent',
          },
          ':focus-visible': {
            borderColor: 'blue.20',
          },
          ':disabled': {
            ':hover': {
              backgroundColor: 'blue.50',
            },
          },
        }
      : {
          backgroundColor: 'blue.50',
          color: 'gray.0',
          ':hover': {
            backgroundColor: 'blue.40',
          },
          ':active': {
            backgroundColor: 'blue.60',
            borderColor: 'transparent',
          },
          ':focus-visible': {
            borderColor: 'blue.70',
          },
          ':disabled': {
            ':hover': {
              backgroundColor: 'blue.50',
            },
          },
        },
  accent: (mode: modes) =>
    mode === modes.dark
      ? {
          backgroundColor: 'mint.50',
          color: 'gray.5',
          ':hover': {
            backgroundColor: 'mint.40',
          },
          ':active': {
            backgroundColor: 'mint.60',
            borderColor: 'transparent',
          },
          ':focus-visible': {
            borderColor: 'blue.20',
          },
          ':disabled': {
            ':hover': {
              backgroundColor: 'mint.50',
            },
          },
        }
      : {
          backgroundColor: 'mint.50',
          color: 'gray.0',
          ':hover': {
            backgroundColor: 'mint.40',
          },
          ':active': {
            backgroundColor: 'mint.60',
            borderColor: 'transparent',
          },
          ':focus-visible': {
            borderColor: 'mint.70',
          },
          ':disabled': {
            ':hover': {
              backgroundColor: 'mint.50',
            },
          },
        },
  danger: (mode: modes) =>
    mode === modes.dark
      ? {
          backgroundColor: 'red.50',
          color: 'gray.5',
          ':hover': {
            backgroundColor: 'red.40',
          },
          ':active': {
            backgroundColor: 'red.60',
            borderColor: 'transparent',
          },
          ':focus-visible': {
            borderColor: 'blue.20',
          },
          ':disabled': {
            ':hover': {
              backgroundColor: 'red.50',
            },
          },
        }
      : {
          backgroundColor: 'red.50',
          color: 'gray.0',
          ':hover': {
            backgroundColor: 'red.40',
          },
          ':active': {
            backgroundColor: 'red.60',
            borderColor: 'transparent',
          },
          ':focus-visible': {
            borderColor: 'red.70',
          },
          ':disabled': {
            ':hover': {
              backgroundColor: 'red.50',
            },
          },
        },
  hollow: (mode: modes) =>
    mode === modes.dark
      ? {
          backgroundColor: 'transparent',
          color: 'gray.5',
          ':hover': {
            backgroundColor: 'gray.60',
          },
          ':active': {
            backgroundColor: 'gray.100',
            borderColor: 'transparent',
          },
          ':focus-visible': {
            borderColor: 'blue.20',
            backgroundColor: 'gray.70',
            ':active': {
              backgroundColor: 'gray.100',
            },
          },
          ':disabled': {
            ':hover': {
              backgroundColor: 'transparent',
            },
          },
        }
      : {
          backgroundColor: 'transparent',
          color: 'gray.80',
          ':hover': {
            backgroundColor: 'gray.2',
          },
          ':active': {
            backgroundColor: 'gray.20',
            borderColor: 'transparent',
          },
          ':focus-visible': {
            borderColor: 'blue.50',
            backgroundColor: 'gray.5',
            ':active': {
              backgroundColor: 'gray.20',
            },
          },
          ':disabled': {
            ':hover': {
              backgroundColor: 'transparent',
            },
          },
        },
};

const baseStyles = themeHelper({
  fontWeight: 'normal',
  fontFamily: 'default',
  fontSize: 3,
  lineHeight: 2,
  borderRadius: 3,
  borderColor: 'transparent',
  borderWidth: '0.125rem',
  borderStyle: 'solid',
  outline: 'none',
  textDecoration: 'none',
  py: 3,
  px: 5,
  ':disabled': {
    opacity: 0.4,
  },
});

const variantStyles = ({ variant = 'primary', ...props }: ButtonProps) => {
  const mode = props.theme?.mode ?? modes.light;
  const variantStyleObject = variants[variant](mode);

  return themeHelper({ ...variantStyleObject })(props);
};

const Button = styled.button<ButtonProps>(baseStyles, variantStyles, sxPropHelper);

export default Button;
