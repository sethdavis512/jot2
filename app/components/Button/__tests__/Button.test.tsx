/**
 * @jest-environment jsdom
 */

import React from 'react';
import { matchers } from '@emotion/jest';
import { ThemeProvider } from '@emotion/react';
import { render, screen } from '@testing-library/react';

import Button from '../';
import theme, { modes } from '../../../utils/theme';

expect.extend(matchers);

describe('Button', () => {
  it('should get theme values for styles', () => {
    render(
      <ThemeProvider theme={{ ...theme, mode: modes.light }}>
        <Button sx={{ color: 'blue.50' }} />
      </ThemeProvider>,
    );

    expect(screen.getByRole('button')).toHaveStyleRule('color', theme.colors.blue[50]);
  });

  it('should have correct styles for primary button with light mode', () => {
    render(
      <ThemeProvider theme={{ ...theme, mode: modes.light }}>
        <Button />
      </ThemeProvider>,
    );

    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('should have correct styles for primary button with dark mode', () => {
    render(
      <ThemeProvider theme={{ ...theme, mode: modes.dark }}>
        <Button />
      </ThemeProvider>,
    );

    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('should have correct styles for standard button with light mode', () => {
    render(
      <ThemeProvider theme={{ ...theme, mode: modes.light }}>
        <Button variant="standard" />
      </ThemeProvider>,
    );

    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('should have correct styles for standard button with dark mode', () => {
    render(
      <ThemeProvider theme={{ ...theme, mode: modes.dark }}>
        <Button variant="standard" />
      </ThemeProvider>,
    );

    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('should have correct styles for danger button with light mode', () => {
    render(
      <ThemeProvider theme={{ ...theme, mode: modes.light }}>
        <Button variant="danger" />
      </ThemeProvider>,
    );

    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('should have correct styles for danger button with dark mode', () => {
    render(
      <ThemeProvider theme={{ ...theme, mode: modes.dark }}>
        <Button variant="danger" />
      </ThemeProvider>,
    );

    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('should have correct styles for accent button with light mode', () => {
    render(
      <ThemeProvider theme={{ ...theme, mode: modes.light }}>
        <Button variant="accent" />
      </ThemeProvider>,
    );

    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('should have correct styles for accent button with dark mode', () => {
    render(
      <ThemeProvider theme={{ ...theme, mode: modes.dark }}>
        <Button variant="accent" />
      </ThemeProvider>,
    );

    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('should have correct styles for hollow button with light mode', () => {
    render(
      <ThemeProvider theme={{ ...theme, mode: modes.light }}>
        <Button variant="hollow" />
      </ThemeProvider>,
    );

    expect(screen.getByRole('button')).toMatchSnapshot();
  });

  it('should have correct styles for hollow button with dark mode', () => {
    render(
      <ThemeProvider theme={{ ...theme, mode: modes.dark }}>
        <Button variant="hollow" />
      </ThemeProvider>,
    );

    expect(screen.getByRole('button')).toMatchSnapshot();
  });
});
