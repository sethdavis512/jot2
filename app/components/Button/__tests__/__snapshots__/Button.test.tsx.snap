// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should have correct styles for accent button with dark mode 1`] = `
.emotion-0 {
  font-weight: 400;
  font-family: "Source Sans Pro","Roboto","Helvetica Neue",Helvetica,Arial,sans-serif;
  font-size: 1rem;
  line-height: 125%;
  border-radius: 0.25rem;
  border-color: transparent;
  border-width: 0.125rem;
  border-style: solid;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background-color: #008480;
  color: #EDEFF0;
}

.emotion-0:disabled {
  opacity: 0.4;
}

.emotion-0:hover {
  background-color: #00A398;
}

.emotion-0:active {
  background-color: #0F6460;
  border-color: transparent;
}

.emotion-0:focus-visible {
  border-color: #A1D3FF;
}

.emotion-0:disabled:hover {
  background-color: #008480;
}

<button
  class="emotion-0"
/>
`;

exports[`Button should have correct styles for accent button with light mode 1`] = `
.emotion-0 {
  font-weight: 400;
  font-family: "Source Sans Pro","Roboto","Helvetica Neue",Helvetica,Arial,sans-serif;
  font-size: 1rem;
  line-height: 125%;
  border-radius: 0.25rem;
  border-color: transparent;
  border-width: 0.125rem;
  border-style: solid;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background-color: #008480;
  color: #FFFFFF;
}

.emotion-0:disabled {
  opacity: 0.4;
}

.emotion-0:hover {
  background-color: #00A398;
}

.emotion-0:active {
  background-color: #0F6460;
  border-color: transparent;
}

.emotion-0:focus-visible {
  border-color: #0B4B3F;
}

.emotion-0:disabled:hover {
  background-color: #008480;
}

<button
  class="emotion-0"
/>
`;

exports[`Button should have correct styles for danger button with dark mode 1`] = `
.emotion-0 {
  font-weight: 400;
  font-family: "Source Sans Pro","Roboto","Helvetica Neue",Helvetica,Arial,sans-serif;
  font-size: 1rem;
  line-height: 125%;
  border-radius: 0.25rem;
  border-color: transparent;
  border-width: 0.125rem;
  border-style: solid;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background-color: #E52207;
  color: #EDEFF0;
}

.emotion-0:disabled {
  opacity: 0.4;
}

.emotion-0:hover {
  background-color: #FB5A47;
}

.emotion-0:active {
  background-color: #B50909;
  border-color: transparent;
}

.emotion-0:focus-visible {
  border-color: #A1D3FF;
}

.emotion-0:disabled:hover {
  background-color: #E52207;
}

<button
  class="emotion-0"
/>
`;

exports[`Button should have correct styles for danger button with light mode 1`] = `
.emotion-0 {
  font-weight: 400;
  font-family: "Source Sans Pro","Roboto","Helvetica Neue",Helvetica,Arial,sans-serif;
  font-size: 1rem;
  line-height: 125%;
  border-radius: 0.25rem;
  border-color: transparent;
  border-width: 0.125rem;
  border-style: solid;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background-color: #E52207;
  color: #FFFFFF;
}

.emotion-0:disabled {
  opacity: 0.4;
}

.emotion-0:hover {
  background-color: #FB5A47;
}

.emotion-0:active {
  background-color: #B50909;
  border-color: transparent;
}

.emotion-0:focus-visible {
  border-color: #8B0A03;
}

.emotion-0:disabled:hover {
  background-color: #E52207;
}

<button
  class="emotion-0"
/>
`;

exports[`Button should have correct styles for hollow button with dark mode 1`] = `
.emotion-0 {
  font-weight: 400;
  font-family: "Source Sans Pro","Roboto","Helvetica Neue",Helvetica,Arial,sans-serif;
  font-size: 1rem;
  line-height: 125%;
  border-radius: 0.25rem;
  border-color: transparent;
  border-width: 0.125rem;
  border-style: solid;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background-color: transparent;
  color: #EDEFF0;
}

.emotion-0:disabled {
  opacity: 0.4;
}

.emotion-0:hover {
  background-color: #565C65;
}

.emotion-0:active {
  background-color: #000000;
  border-color: transparent;
}

.emotion-0:focus-visible {
  border-color: #A1D3FF;
  background-color: #3D4551;
}

.emotion-0:focus-visible:active {
  background-color: #000000;
}

.emotion-0:disabled:hover {
  background-color: transparent;
}

<button
  class="emotion-0"
/>
`;

exports[`Button should have correct styles for hollow button with light mode 1`] = `
.emotion-0 {
  font-weight: 400;
  font-family: "Source Sans Pro","Roboto","Helvetica Neue",Helvetica,Arial,sans-serif;
  font-size: 1rem;
  line-height: 125%;
  border-radius: 0.25rem;
  border-color: transparent;
  border-width: 0.125rem;
  border-style: solid;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background-color: transparent;
  color: #2D2E2F;
}

.emotion-0:disabled {
  opacity: 0.4;
}

.emotion-0:hover {
  background-color: #F7F9FA;
}

.emotion-0:active {
  background-color: #C6CACE;
  border-color: transparent;
}

.emotion-0:focus-visible {
  border-color: #0076DA;
  background-color: #EDEFF0;
}

.emotion-0:focus-visible:active {
  background-color: #C6CACE;
}

.emotion-0:disabled:hover {
  background-color: transparent;
}

<button
  class="emotion-0"
/>
`;

exports[`Button should have correct styles for primary button with dark mode 1`] = `
.emotion-0 {
  font-weight: 400;
  font-family: "Source Sans Pro","Roboto","Helvetica Neue",Helvetica,Arial,sans-serif;
  font-size: 1rem;
  line-height: 125%;
  border-radius: 0.25rem;
  border-color: transparent;
  border-width: 0.125rem;
  border-style: solid;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background-color: #0076DA;
  color: #EDEFF0;
}

.emotion-0:disabled {
  opacity: 0.4;
}

.emotion-0:hover {
  background-color: #2491FF;
}

.emotion-0:active {
  background-color: #005EA2;
  border-color: transparent;
}

.emotion-0:focus-visible {
  border-color: #A1D3FF;
}

.emotion-0:disabled:hover {
  background-color: #0076DA;
}

<button
  class="emotion-0"
/>
`;

exports[`Button should have correct styles for primary button with light mode 1`] = `
.emotion-0 {
  font-weight: 400;
  font-family: "Source Sans Pro","Roboto","Helvetica Neue",Helvetica,Arial,sans-serif;
  font-size: 1rem;
  line-height: 125%;
  border-radius: 0.25rem;
  border-color: transparent;
  border-width: 0.125rem;
  border-style: solid;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background-color: #0076DA;
  color: #FFFFFF;
}

.emotion-0:disabled {
  opacity: 0.4;
}

.emotion-0:hover {
  background-color: #2491FF;
}

.emotion-0:active {
  background-color: #005EA2;
  border-color: transparent;
}

.emotion-0:focus-visible {
  border-color: #0B4778;
}

.emotion-0:disabled:hover {
  background-color: #0076DA;
}

<button
  class="emotion-0"
/>
`;

exports[`Button should have correct styles for standard button with dark mode 1`] = `
.emotion-0 {
  font-weight: 400;
  font-family: "Source Sans Pro","Roboto","Helvetica Neue",Helvetica,Arial,sans-serif;
  font-size: 1rem;
  line-height: 125%;
  border-radius: 0.25rem;
  border-color: transparent;
  border-width: 0.125rem;
  border-style: solid;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background-color: #3D4551;
  color: #EDEFF0;
}

.emotion-0:disabled {
  opacity: 0.4;
}

.emotion-0:hover {
  background-color: #565C65;
}

.emotion-0:active {
  background-color: #000000;
  border-color: transparent;
}

.emotion-0:focus-visible {
  border-color: #A1D3FF;
}

.emotion-0:disabled:hover {
  background-color: #3D4551;
}

<button
  class="emotion-0"
/>
`;

exports[`Button should have correct styles for standard button with light mode 1`] = `
.emotion-0 {
  font-weight: 400;
  font-family: "Source Sans Pro","Roboto","Helvetica Neue",Helvetica,Arial,sans-serif;
  font-size: 1rem;
  line-height: 125%;
  border-radius: 0.25rem;
  border-color: transparent;
  border-width: 0.125rem;
  border-style: solid;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  background-color: #EDEFF0;
  color: #2D2E2F;
}

.emotion-0:disabled {
  opacity: 0.4;
}

.emotion-0:hover {
  background-color: #F7F9FA;
}

.emotion-0:active {
  background-color: #C6CACE;
  border-color: transparent;
}

.emotion-0:focus-visible {
  border-color: #0076DA;
}

.emotion-0:disabled:hover {
  background-color: #EDEFF0;
}

<button
  class="emotion-0"
/>
`;
